astropy/units/structured.py

```diff
@@ -0,0 +1,465 @@
+                names = _names_from_dtype(names)
+            else:
+                if not isinstance(names, tuple):
+                    names = (names,)
+                names = _normalize_names(names)
+
+        if not isinstance(units, tuple):
+            units = Unit(units)
+            if isinstance(units, StructuredUnit):
+                # Avoid constructing a new StructuredUnit if no field names
+                # are given, or if all field names are the same already anyway.
+                if names is None or units.field_names == names:
+                    return units
+
+                # Otherwise, turn (the upper level) into a tuple, for renaming.
+                units = units.values()
+            else:
+                # Single regular unit: make a tuple for iteration below.
+                units = (units,)
+
+        if names is None:
+            names = tuple(f'f{i}' for i in range(len(units)))
+
+        elif len(units) != len(names):
+            raise ValueError("lengths of units and field names must match.")
+
+        converted = []
+        for unit, name in zip(units, names):
+            if isinstance(name, list):
+                # For list, the first item is the name of our level,
+                # and the second another tuple of names, i.e., we recurse.
+                unit = cls(unit, name[1])
+                name = name[0]
+            else:
+                # We are at the lowest level.  Check unit.
+                unit = Unit(unit)
+                if dtype is not None and isinstance(unit, StructuredUnit):
+                    raise ValueError("units do not match in depth with field "
+                                     "names from dtype or structured unit.")
+
+            converted.append(unit)
+
+        self = super().__new__(cls)
+        if dtype is None:
+            dtype = np.dtype([((name[0] if isinstance(name, list) else name),
+                               DTYPE_OBJECT) for name in names])
+        # Decay array to void so we can access by field name and number.
+        self._units = np.array(tuple(converted), dtype)[()]
+        return self
+
+    @property
+    def field_names(self):
+        """Possibly nested tuple of the field names of the parts."""
+        return tuple(([name, unit.field_names]
+                      if isinstance(unit, StructuredUnit) else name)
+                     for name, unit in self.items())
+
+    # Allow StructuredUnit to be treated as an (ordered) mapping.
+    def __len__(self):
+        return len(self._units.dtype.names)
+
+    def __getitem__(self, item):
+        # Since we are based on np.void, indexing by field number works too.
+        return self._units[item]
+
+    def values(self):
+        return self._units.item()
+
+    def keys(self):
+        return self._units.dtype.names
+
+    def items(self):
+        return zip(self._units.dtype.names, self._units.item())
```

Comment:
Should this use [``ItemsView``](https://docs.python.org/3/library/collections.abc.html#collections.abc.ItemsView)

Reply:
And ``zip`` doesn't look good when printed. Maybe consume the iterator into a tuple?

Reply:
Since both the names and values are tuples and not very long, I agree one might as well make it a tuple. I guess the only downside is that it will slow down some of the code which uses `.items()` as an iterator, but the additional 100 ns is probably negligible to everything else...

Reply:
This is a frequently encountered problem Python really need more sophisticated tools for building views. ValuesView, KeysView, and ItemsView should make dict-like views.

Reply:
Would be nice for `repr`, though I am still quite enchanted with the idea that all you need to know about the output of `.items()` is that it is an iterable that provides key, value sequences. (If you hadn't noticed already, I love ducktyping!)